include/gmsh.tcc

```diff
@@ -4,137 +4,151 @@
-      std::istringstream istream(line);
-
-      if (line.find('#') == std::string::npos && line != "") {
-        //! Coordinates of vertex
-        std::array<double, Tdim> vertex;
-
-        istream >> vertid;
-        istream >> vertex.at(0) >> vertex.at(1) >> vertex.at(2);
-        vertices_.insert(std::make_pair(vertid, vertex));
-      }
+      istream >> vertid;
+      istream >> vertex.at(0) >> vertex.at(1) >> vertex.at(2);
+      this->vertices_.insert(std::make_pair(vertid, vertex));
     }
-    infile.close();
   }
-  std::cout << "Number of Vertices: " << vertices_.size() << '\n';
+
+  this->nvertices_ = vertices_.size();
+
+  //! Check that the number of vertices are correct
+  if (nvertices_ != nvertices) {
+    std::cout << "Error in the code, number of vertices do not match.\n";
+  }
+
+  std::cout << "Number of Vertices: " << nvertices_ << '\n';
 }
 
 //! Read GMSH elements
 //! \tparam Tdim Dimension
 //! \tparam Tvertices Number of vertices in element
 //! \param[in] filename Input mesh filename
 template <unsigned Tdim, unsigned Tvertices>
-void GMSH<Tdim, Tvertices>::read_elements(const std::string& filename) {
+void GMSH<Tdim, Tvertices>::read_elements(std::ifstream& file) {
+
+  //! Find the line of interest
+  read_keyword(file, "$Elements");
+
+  std::string line;
+  std::getline(file, line);
+  std::istringstream istream(line);
 
-  //! Number of vertices
-  double nvertices = std::numeric_limits<double>::max();
   //! Number of elements
-  double nelements = std::numeric_limits<double>::max();
+  unsigned nelements = std::numeric_limits<unsigned>::max();
+  istream >> nelements;
+  getline(istream, line);
+
   //! Element type
   double elementtype = std::numeric_limits<double>::max();
 
-  double physical = std::numeric_limits<double>::max();
-  double elementry = std::numeric_limits<double>::max();
   //! Element id
   unsigned elementid = std::numeric_limits<unsigned>::max();
 
-  const unsigned toplines = 4;
-  // specify element type 4 = tetrahedral
-  const unsigned element_type = 4;
+  double physical = std::numeric_limits<double>::max();
+
+  double elementry = std::numeric_limits<double>::max();
 
   //! Array to store vertices coordinates
   std::array<double, Tvertices> elementarray;
 
-  std::fstream infile;
-  infile.open(filename, std::ios::in);
-
-  //! Check if input msh file is good
-  if (infile.good()) {
-    std::cout << "Element file found" << '\n';
-
-    std::string line;
-
-    //! Ignore first 4 lines
-    for (unsigned k = 0; k < toplines; ++k) {
-      std::getline(infile, line);
-    }
-    //! Get number of vertices
-    infile >> nvertices;
-    getline(infile, line);
-    //! Skip past vertices section
-    for (int i = 0; i < nvertices; ++i) {
-      std::getline(infile, line);
-    }
-    //! Skip past element headers
-    for (unsigned l = 0; l < 2; ++l) {
-      std::getline(infile, line);
-    }
+  //! specify element type 4 = tetrahedral
+  const unsigned element_type = 4;
```

Comment:
I thought we use `quadrilateral` or `hexahedron`.

Reply:
in `mesh` file, we use number 4 or 8 I think

Reply:
yes, but `tetraheron` is not an element we use.

Reply:
Understood!